### PR TITLE
Refactoring and cleaning the type enumerator for sets

### PR DESCRIPTION
**Pull request by mudathirmahgoub**
This PR rewrites set enumerator and changes the order of enumerated sets. The PR simplifies the code and adds unit tests to verify the correctness. Before approving, please make sure the unit tests cover all the edge cases.

### CODE REVIEW

File: src/theory/sets/theory_sets_type_enumerator.h [hunk 1]
```diff
@@ -2,185 +2,164 @@
...
+                           << d_currentSet << std::endl;
 
-    Assert(n.isConst());
-    Assert(n == Rewriter::rewrite(n));
-
-    return n;
+    return d_currentSet;
   }
 
+  /**
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_type_enumerator.h]: Can you move this (and other functions) to a *.cpp file?

Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_type_enumerator.h]: done

File: src/theory/sets/theory_sets_type_enumerator.cpp [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                           << std::endl;
+    Trace("set-type-enum") << "SetEnumerator::operator++ d_currentSet = "
+                           << d_currentSet << std::endl;
+    return *this;
+  }
+
+  d_currentSetIndex++;
+
+  // if the index is a power of 2, get a new element from d_elementEnumerator
+  if (d_currentSetIndex == (unsigned)(1 << d_elementsSoFar.size()))
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_type_enumerator.cpp]:
```suggestion
  if (d_currentSetIndex == static_cast<unsigned>(1 << d_elementsSoFar.size()))
```

File: src/theory/sets/theory_sets_type_enumerator.h [hunk 3]
```diff
@@ -2,185 +2,66 @@
...
+   * following sequence consisting of 16 sets:
+   * {}, {00}, {01}, {00, 01}, {10}, {00, 10}, {01, 10}, {00, 01, 10}, ...,
+   * {00, 01, 10, 11}
+   */
+  SetEnumerator& operator++() override;
+
+  bool isFinished() override;
+
+ private:
+  NodeManager* d_nodeManager;
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_type_enumerator.h]: Can you add a one line comment for each of these?

File: src/theory/sets/theory_sets_type_enumerator.cpp [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+      Trace("set-type-enum")
+          << "SetEnumerator::operator++ d_currentSet = " << d_currentSet
+          << std::endl;
+      return *this;
+    }
+
+    // get a new element and return it as a singleton set
+    Node element = *d_elementEnumerator;
+    d_elementsSoFar.push_back(element);
+    d_currentSet = d_nodeManager->mkNode(Kind::SINGLETON, element);
```
Comment (ajreynol) [inline src/theory/sets/theory_sets_type_enumerator.cpp]:
```suggestion
    d_currentSet = d_nodeManager->mkNode(kind::SINGLETON, element);
```

File: src/theory/sets/theory_sets_type_enumerator.cpp (hunk 2 above)
Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_type_enumerator.cpp]: done

File: src/theory/sets/theory_sets_type_enumerator.h (hunk 3 above)
Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_type_enumerator.h]: done

File: src/theory/sets/theory_sets_type_enumerator.cpp (hunk 4 above)
Comment (mudathirmahgoub) [reply src/theory/sets/theory_sets_type_enumerator.cpp]: Done. wondering what is the difference between `Kind` and `kind`?

**Comment by ajreynol** [comment]
Thanks a lot for doing this, looks way better now.
